veps/vep-4.md

```diff
@@ -0,0 +1,342 @@
+
+For currently-exported variables, this is trickier. My proposal:
+
+Currently:
+```go
+// go/vt/vttablet/tmclient/client.go
+package tmclient
+
+import "flag"
+
+var TabletManagerProtocol = flag.String(...)
+```
+
+After:
+```go
+// go/vt/vttablet/tmclient/client.go
+package tmclient
+
+import flag "github.com/spf13/pflag"
+
+var clientProtocol string
+
+func AddFlags(fs *flag.FlagSet) {
+    fs.StringVar(&clientProtocol, "tablet_manager_client_protocol", ...)
+}
+
+func TabletManagerProtocol() string {
+    return clientProtocol
+}
+
+// mypkg/some_test.go
+package mypkg
+
+import (
+    "testing"
+
+    flag "github.com/spf13/pflag"
+)
+
+func TestMyFunc(t *testing.T) {
+    ...
+    flag.Set("tablet_manager_client_protocol", "mytestvalue")
+}
+```
+
+Commentary:
+* This makes flag variables read-only ... ish.
+The only way to set `clientProtocol` outside of `package tmclient` is with a call to `pflag.Set`.
+This is easy to lint, and we can enforce it by convention in code review, or we can add a CI hook if it becomes frequent enough to warrant the extra effort.
+* It does **not** make flag-setting in tests threadsafe.
+This is a bummer to me, as it reduces the number of tests we can mark as `t.Parallel()`, but that's not one of our goals, so figuring that out can wait for later and should not stop us from making this clear improvement to The Flag Situation.
+* It allows us to import packages without polluting help text with flags if we just need one or two utility functions.
+* As we move each non-`cmd` package to this form, we will need to audit each `cmd`, determine if they need those flags, and add the appropriate call to `AddFlags`.
+* Packages that depend on an appropriate default value need to duplicate that default in the flag definition and the variable declaration ([c.f.](https://play.golang.org/p/aWPA-IgiGq_r)).
+For example, it should be:
+```go
+package topo
+import (
+    "time"
+
+    flag "github.com/spf13/pflag"
+)
+
+var remoteOperationTimeout = time.Second*30
+func AddFlags(fs *flag.FlagSet) { fs.DurationVar(&remoteOperationTimeout, "remote_operation_timeout" time.Second*30, "...") }
+```
+
+and not:
+```go
+package topo
+
+import (
+    "time"
+
+    flag "github.com/spf13/pflag"
+)
+
+var remoteOperationTimeout time.Duration
+func AddFlags(fs *flag.FlagSet) { fs.DurationVar(&remoteOperationTimeout, "remote_operation_timeout", time.Second*30, "...") }
+```
+
+### Phase 4: `viper` and beyond
+
+At this point, we're using `pflag` everywhere, and our flags are well-encapsulated within the packages that define them.
+However, there's more we can improve!
+
+For binaries that have many options (like `vttablet`), specifying all of those on the command-line is cumbersome.
+In addition, we currently don't support live-reloading of configuration variables.
+Live-reload is a huge usability win; consider for example being able to adjust an operation timeout without having to restart the component ... :chefs-kiss:
```

Comment:
Agreed. We will still need to cautious about which flags can actually take effect on a running instance. What if someone wants to change a connection pool size, can we actually support that?

Reply:
Right, I think that's a separate design/rfc/project (i.e. one for "tablet connpool live reload" and another for "vreplication timeouts live reload" and so on). Viper only makes this possible/easy to implement, but we still have to add the watch call and reload logic (which is necessarily app-specific and therefore not provided by viper) ourselves explicitly. I will add some language to clarify that here!